File: src/main/java/com/gary/GalacticTrading/io/OutputProcessor.java
package com.gary.GalacticTrading.io;

import com.gary.GalacticTrading.calculator.MetalAndMultipleCalculator;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.Resource;
import org.springframework.core.io.ResourceLoader;
import org.springframework.stereotype.Service;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;

@Service
@Slf4j
@RequiredArgsConstructor
public class OutputProcessor {
    private final MetalAndMultipleCalculator metalAndMultipleCalculator;
    private final ResourceLoader resourceLoader;
    private List<String> contents = new ArrayList<>();

    public void saveToOutputFile(final String[] unitsAndMetalQuery) {
        log.debug("Saving to buffer and ready for output file...");

            if (unitsAndMetalQuery.length == 1) {
                contents.add(unitsAndMetalQuery[0]);
                return;
            }

            String temp = String.join(" ", unitsAndMetalQuery);
            temp += " is ";
            final int length = unitsAndMetalQuery.length;
            if (metalAndMultipleCalculator.getMetalNameMultiplerMap().get(unitsAndMetalQuery[length - 1]) != null) {
                final String metalName = unitsAndMetalQuery[length - 1];
                final String interGalacticUnits = String.join(" ", unitsAndMetalQuery).replace(metalName, "");
                temp += String.valueOf(
                        metalAndMultipleCalculator.calculateMetalValue(interGalacticUnits, metalName));
                temp += " Credits";
            } else {
                temp += String.valueOf(metalAndMultipleCalculator.calculateMetalValue(
                        String.join(" ", unitsAndMetalQuery), null));
            }
            contents.add(temp);
    }

    public void writeToFile() {
        Resource resource =  resourceLoader.getResource("classpath:output.txt");
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(resource.getFile()))) {
            for (String str : contents) {
                writer.write(str + System.lineSeparator());
            }
        } catch (IOException e) {
            log.error("Error: Unable to write to file output.txt");
        }
    }

}

File: src/main/kotlin/com/gary/GalacticTrading/io/OutputProcessor.kt
package com.gary.GalacticTrading.io

import com.gary.GalacticTrading.calculator.MetalAndMultipleCalculator
import mu.KotlinLogging
import org.springframework.stereotype.Service
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException

@Service
data class OutputProcessor(private val metalAndMultipleCalculator: MetalAndMultipleCalculator) {
    var contents: List<String> = ArrayList()
    private val log = KotlinLogging.logger {}

    fun saveForOutput(unitsAndMetalQuery: Array<String>) {
        log.debug("Saving to buffer and ready for output file...")

        if (unitsAndMetalQuery.size == 1) {
            contents += unitsAndMetalQuery[0]
            return
        }

        var temp = unitsAndMetalQuery.joinToString(" ")
        temp += " is "
        val length = unitsAndMetalQuery.size
        if (metalAndMultipleCalculator.metalNameMultiplerMap[unitsAndMetalQuery[length - 1]] != null) {
            val metalName = unitsAndMetalQuery[length - 1]
            val interGalacticUnits = unitsAndMetalQuery.joinToString(" ").replace(metalName, "")
            temp += metalAndMultipleCalculator.calculateMetalValue(interGalacticUnits, metalName).toString()
            temp += " Credits"
        } else {
            temp += metalAndMultipleCalculator.calculateMetalValue(
                unitsAndMetalQuery.joinToString(" "), ""
            ).toString()
        }
        contents += temp
    }

    fun writeToFile(outputFileName: String) {
        val resourcesDir = File("src/main/resources/")
        val file = File(resourcesDir.absolutePath + File.separator + outputFileName)
        if (file.createNewFile()) {
            log.info("File created: " + file.name)
        } else {
            log.info("File already exists.")
        }

        try {
            BufferedWriter(FileWriter(file)).use { writer ->
                for (str in contents) {
                    writer.write(str + System.lineSeparator())
                }
                log.info("Output written to file: {}", file.absolutePath)
                contents = ArrayList()
            }
        } catch (e: IOException) {
            log.error("Error: Unable to write to file output.txt")
            throw e
        }
    }

    fun reset() {
        contents = ArrayList()
    }
}
